### PR TITLE
BiMPM model

### PR DESCRIPTION
**Pull request by handsomezebra**
This PR includes the implementation of BiMPM model (#1503) and a Quora paraphrase dataset reader. I am still working on some comments improvements and memory usage optimization but I want to send out the PR to collect feedback too.

Thanks!

### CODE REVIEW

**Comment by handsomezebra** [comment]
Looks like documents for the classes are missing. I will update the PR.

**Comment by handsomezebra** [comment]
Pinging @schmmd , please help to provide comments to this pack. 
I asked the original author of the paper https://github.com/zhiguowang/ to review the code he said he is busy recently and will review once he get some time.

**Comment by matt-gardner** [comment]
Sorry to be slow @handsomezebra, we had a hackathon last week at AI2 that put a bunch of other stuff on pause.  I'll look at this, though today is pretty busy for me, and it'll probably be tomorrow before I get to it.

File: allennlp/data/dataset_readers/quora_paraphrase.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        super().__init__(lazy)
+        self._tokenizer = tokenizer or WordTokenizer(JustSpacesWordSplitter())
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+    @overrides
+    def _read(self, file_path):
+        logger.info("Reading instances from lines in file at: %s", file_path)
+        file_name, member = parse_file_uri(file_path)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/quora_paraphrase.py]: You've duplicated the logic below between the two cases (where `member is None` and where it isn't).  Seems like it would be cleaner to push the file decision logic into this method.  So you'd have something like:

```python
def open_uri(file_path):
    file_name, member = parse_file_uri(file_path)
    file_path = cached_path(file_path)
    if member is None:
        return open(file_path, 'r')
    else:
        # zipfile opening logic here

def _read(self, file_path):
    with open_uri(file_path) as data_file:
       tsvin = ...
```

@joelgrus probably knows more than I do about how to handle the `with` stuff properly in this case.

File: allennlp/data/dataset_readers/quora_paraphrase.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    match = re.fullmatch(r'\((.*)\)#(.*)', uri)      # pylint: disable=anomalous-backslash-in-string
+    if match and len(match.groups()) == 2:
+        return match.groups()[0], match.groups()[1]
+    else:
+        return uri, None
+
+@DatasetReader.register("quora_paraphrase")
+class QuoraParaphraseDatasetReader(DatasetReader):
+    """
+    Reads a Quora paraphrase data
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/quora_paraphrase.py]: An explanation of the expected data format here would be nice (e.g., this is a TSV file, with the following columns).

File: allennlp/models/bimpm.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        Second encoder layer for the premise and hypothesis
+    matcher_fw2 : ``BiMPMMatching``
+        BiMPM matching for the forward output of second encoder layer
+    matcher_bw2 : ``BiMPMMatching``
+        BiMPM matching for the backward output of second encoder layer
+    aggregator : ``Seq2VecEncoder``
+        Aggregator of all BiMPM matching vectors
+    classifier_feedforward : ``FeedForward``
+        Fully connected layers for classification.
+    dropout : ``float``
```
Comment (matt-gardner) [inline allennlp/models/bimpm.py]: This should be marked optional.

File: allennlp/models/bimpm.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    ----------
+    vocab : ``Vocabulary``
+    text_field_embedder : ``TextFieldEmbedder``
+        Used to embed the ``premise`` and ``hypothesis`` ``TextFields`` we get as input to the
+        model.
+    word_matcher : ``BiMPMMatching``
+        BiMPM matching on the output of word embeddings of premise and hypothesis.
+    encoder1 : ``Seq2SeqEncoder``
+        First encoder layer for the premise and hypothesis
+    matcher_fw1 : ``BiMPMMatching``
```
Comment (matt-gardner) [inline allennlp/models/bimpm.py]: Instead of `fw` and `bw`, can you expand those to `forward` and `backward`?  It's much easier to read.

File: allennlp/models/bimpm.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        self.matcher_fw2 = matcher_fw2
+        self.matcher_bw2 = matcher_bw2
+
+        self.aggregator = aggregator
+
+        matching_dim = self.word_matcher.get_output_dim() + \
+                       self.matcher_fw1.get_output_dim() + self.matcher_bw1.get_output_dim() + \
+                       self.matcher_fw2.get_output_dim() + self.matcher_bw2.get_output_dim()
+
+        if matching_dim != self.aggregator.get_input_dim():
```
Comment (matt-gardner) [inline allennlp/models/bimpm.py]: We have a `check_dimensions_match` function that makes these checks a little easier: https://github.com/allenai/allennlp/blob/580dc8b0e2c6491d4d75b54c3b15b34b462e0c67/allennlp/models/esim.py#L82-L83

File: allennlp/modules/bimpm_matching.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    keepdim : ``bool``
+        Whether to keep dimension
+    min_val : ``float``
+        The minimal value for paddings
+
+    Returns
+    -------
+    A ``torch.Tensor`` of including the maximum values.
+    """
+    replaced_vector = vector + (1.0 - mask) * min_val
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: You can use `vector.masked_fill_()` here instead of this logic.

File: allennlp/modules/bimpm_matching.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                dim: int,
+                keepdim: bool = False,
+                eps: float = 1e-8) -> torch.Tensor:
+    """
+    To calculate mean along certain dimensions on masked values
+
+    Parameters
+    ----------
+    vector : ``torch.Tensor``
+        The vector to calculate mean, assume unmasked parts are already zeros
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: For a general utility, I wouldn't make this assumption, instead doing a `vector.masked_fill_(0, mask)` before the `torch.sum()`.

File: allennlp/modules/bimpm_matching.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.registrable import FromParams
+from allennlp.nn.util import get_lengths_from_binary_sequence_mask
+
+
+def masked_max(vector: torch.Tensor,
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: These two functions should go in `nn.util`.

File: allennlp/modules/bimpm_matching.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    is_forward : ``bool``, required
+        Whether the matching is for forward sequence or backward sequence
+    hidden_dim : ``int``, optional (default = 100)
+        The hidden dimension of the representations
+    num_perspective : ``int``, optional (default = 20)
+        The number of perspective for matching
+    share_weight_between_directions : ``bool``, optional (default = True)
+        If True, share weight between premise to hypothesis and hypothesisto premise,
+        useful for non-symmetric tasks
+    wo_full_match : ``bool``, optional (default = False)
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: What is `wo`?  "without"?  That's not obvious.

File: allennlp/modules/bimpm_matching.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        else:
+            num_matching = int(not wo_full_match) * 2 + \
+                           int(not wo_maxpool_match) + \
+                           int(not wo_attentive_match) * 2 + \
+                           int(not wo_max_attentive_match) * 2
+
+        if num_matching <= 0:
+            raise ConfigurationError("At least one of the matching method should be enabled")
+
+        params = [nn.Parameter(torch.rand(num_perspective, hidden_dim)) for i in range(num_matching)]
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: `torch.rand()` isn't a great way to initialize parameters.  Can you use something more reasonable as a default here, like `torch.nn.init.xavier_uniform_`?

Also, having a list of parameters here is pretty hard to interpret if you're looking at parameter values in tensorboard, or during debugging, or similar.  I think it'd be better to give these parameters names, like `full_match_weights`, and assign them directly to `self`, assigning `None` if they aren't selected.  That would also let you get rid of `num_matching`, `mv_idx`, and `mv_idx_increment` entirely, which are a little confusing.

File: allennlp/modules/bimpm_matching.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+    # (batch, num_perspective, seq_len1, seq_len2)
+    mul_result = torch.matmul(vector1, vector2.transpose(2, 3))
+    norm_value = v1_norm * v2_norm.transpose(2, 3)
+
+    # (batch, seq_len1, seq_len2, num_perspective)
+    return (mul_result / norm_value.clamp(min=eps)).permute(0, 2, 3, 1)
+
+
+class BiMPMMatching(nn.Module, FromParams):
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: We prefer to use [Google's definition of camel case](https://github.com/allenai/allennlp/blob/master/STYLE.md#naming), where this would be `BiMpmMatching` (and `BiMpm` instead of `BiMPM`).

File: allennlp/modules/bimpm_matching.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    Returns
+    -------
+    A ``torch.Tensor`` of including the mean values.
+    """
+    value_sum = torch.sum(vector, dim=dim, keepdim=keepdim)
+    value_count = torch.sum(mask, dim=dim, keepdim=keepdim)
+    return value_sum / value_count.clamp(min=eps)
+
+
+def mpm(vector1: torch.Tensor,
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: A more descriptive name would be better here (and below), like `multi_perspective_match()`.

**Comment by handsomezebra** [comment]
Thanks a lot, Matt. I agree on most of the comments and I am making changes now...

File: allennlp/data/dataset_readers/quora_paraphrase.py (hunk 1 above)
Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: @matt-gardner @joelgrus Sure I will try to refactor this. One related question is I am currently using my personal s3 to host the data file `https://s3-us-west-1.amazonaws.com/handsomezebra/public/Quora_question_pair_partition.zip`. Do you think you can help me upload the files to allennlp's s3?

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Yes, I'd be happy to put that data file (and any model files you have) into our s3 bucket. Just tell me what all of them are.

File: allennlp/modules/bimpm_matching.py (hunk 10 above)
Comment (matt-gardner) [reply allennlp/modules/bimpm_matching.py]: Yeah, I know you re-initialize them in your config, but the default should also do something reasonable, in case people forget or don't want to use a special initialization.

File: allennlp/data/dataset_readers/quora_paraphrase.py (hunk 1 above)
Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Just this data file for now https://s3-us-west-1.amazonaws.com/handsomezebra/public/Quora_question_pair_partition.zip
Thanks!

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Ok, I unzipped it and put the train, dev and test files here: `https://s3-us-west-1.amazonaws.com/allennlp/datasets/quora-question-paraphrase/{train,dev,test}.tsv`, along with the README.  I left out the word vector file.

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: (Unzipping it also makes it so you can just remove the zip file logic and simplify that piece of the code.  I'd vote for handling zip files in a more general way in our `common` module, which we can certainly do, but is probably more appropriate for a separate PR.)

Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Sure I will remove the zip file reading. 
I tried the s3 URL https://s3-us-west-1.amazonaws.com/allennlp/datasets/quora-question-paraphrase/train.tsv but got a redirection error. But this URL is fine: https://allennlp.s3-us-west-2.amazonaws.com/datasets/quora-question-paraphrase/train.tsv. I can use the second URL but other data on allennlp doesn't seem to have this problem.

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Ah, sorry, that's my fault, I had a copy-paste error in the link I gave you.

**Comment by handsomezebra** [comment]
Pack updated with all those issues fixed according to the comments.

File: allennlp/data/dataset_readers/quora_paraphrase.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+from allennlp.data.fields import LabelField, TextField, Field
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Tokenizer, WordTokenizer
+from allennlp.data.tokenizers.word_splitter import JustSpacesWordSplitter
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def parse_file_uri(uri: str):
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/quora_paraphrase.py]: I think you don't need this anymore.  (Though, as I said before, I think it's totally reasonable to talk about adding this functionality more generally, in `common.file_utils`.  It should just be in a separate PR.)

File: allennlp/data/dataset_readers/quora_paraphrase.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+                    yield self.text_to_instance(premise=row[1], hypothesis=row[2], label=row[0])
+
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         premise: str,
+                         hypothesis: str,
+                         label: str = None) -> Instance:
+        # pylint: disable=arguments-differ
+        fields: Dict[str, Field] = {}
+        tokenized_premise = self._tokenizer.tokenize(premise)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/quora_paraphrase.py]: Oh, these are pre-tokenized!  That's an important point, thanks for adding it to the class docstring.  I think that means you should _not_ tokenize the text here.  The reason is because in a demo, you'll want to do tokenization (because someone typing text into a demo won't necessarily tokenize it nicely for you).  So if/when we build a demo for this, we'll want the `Predictor` to have a tokenizer and pass pre-tokenized text into here.  So I'd recommend having `premise: List[str]` and `hypothesis: List[str]`, removing the tokenizer, and just calling `row[1].split()` and `row[2].split()` above.  Does that make sense?

File: allennlp/modules/bimpm_matching.py [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+    Multi-Perspective Matching for Natural Language Sentences <https://arxiv.org/abs/1702.03814>`_
+    by Zhiguo Wang et al., 2017.
+    Also please refer to the `TensorFlow implementation <https://github.com/zhiguowang/BiMPM/>`_ and
+    `PyTorch implementation <https://github.com/galsang/BIMPM-pytorch>`_.
+
+    Parameters
+    ----------
+    hidden_dim : ``int``, optional (default = 100)
+        The hidden dimension of the representations
+    num_perspective : ``int``, optional (default = 20)
```
Comment (matt-gardner) [inline allennlp/modules/bimpm_matching.py]: Nit: this should probably be `num_perspectives`, here, in the description on the next line, and as a variable name in the code..

File: allennlp/data/dataset_readers/quora_paraphrase.py (hunk 13 above)
Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Oh, right. Let me remove this.

File: allennlp/modules/bimpm_matching.py (hunk 15 above)
Comment (handsomezebra) [reply allennlp/modules/bimpm_matching.py]: Sure, I will rename it.

**Comment by handsomezebra** [comment]
Currently I have a running experiment on this version of code. It is not finished yet but it's very probable I get 0.89+ accuracy on the Quora test data. After the PR is merged I can test and upload the model.

And yes I am willing to build a demo from it. Let me work on a PR maybe next week.

File: allennlp/data/dataset_readers/quora_paraphrase.py (hunk 14 above)
Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Sure it makes sense. But as text_to_instance is an overridden method, will the change of types cause any side effect? How about we do this in the next PR where we add the predictor and demo?

Comment (handsomezebra) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Also, the raw Quora data is actually not pre-tokenized. Only this particular train/validate/test split is pre-tokenized. At sometime I will want to use the same split but use non-tokenized version to see if better tokenization could further improve the performance.

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Oh, ok, that's a good point.  And yeah, it's fine to worry about this in a subsequent PR.  It's just something to be careful about when this actually goes into a demo.

Comment (matt-gardner) [reply allennlp/data/dataset_readers/quora_paraphrase.py]: Oh, and for the types, we're already ignoring types on these methods, because none of them agree with the super class.  The point of that method existing on the super class is just to tell you that you really should implement a method with this name, though the particulars of how it looks are going to be specific to your data.

**Comment by matt-gardner** [comment]
Awesome, thanks again for all of this work!

**Comment by handsomezebra** [comment]
I put the trained model file here: https://s3-us-west-1.amazonaws.com/handsomezebra/public/bimpm-quora-2018.08.17.tar.gz

Running the following commands under the latest master.
```
./bin/allennlp evaluate https://s3-us-west-1.amazonaws.com/handsomezebra/public/bimpm-quora-2018.08.17.tar.gz https://s3-us-west-2.amazonaws.com/allennlp/datasets/quora-question-paraphrase/test.tsv
```
Got accuracy 89.3%.

**Comment by saurabhvyas** [comment]
Hi, Thanks for your hard work, Can you update the docs, as to how to train your model on my custom data ?

**Comment by handsomezebra** [comment]
@saurabhvyas 
To train and evaluate bimpm model, use the following commands.
```
allennlp train training_config/bimpm_snli.json -s <serialization_folder>
allennlp evaluate <serialization_folder>/model.tar.gz https://s3-us-west-2.amazonaws.com/allennlp/datasets/quora-question-paraphrase/test.tsv
```

If you want to use your custom data, please just change train_data_path and validation_data_path, if your data have the exact same format as Quora's data (tsv file, four columns with label, sentence1, sentence2 and id, word pretokenized).

Otherwise if your data have different format, please write your own dataset reader. Please refer to allennlp's tutorial https://github.com/allenai/allennlp/tree/master/tutorials for some more information.

**Comment by schmmd** [comment]
@handsomezebra yes thank you!  And sorry for the slow responses--I was out last week.  If you're willing to build a demo it would be great to have one for the Quora dataset.  If you make a rough demo we're able to pretty it up.

**Comment by schmmd** [comment]
I moved the model into our S3.

```
$ allennlp evaluate https://s3-us-west-2.amazonaws.com/allennlp/models/bimpm-quora-2018.08.17.tar.gz https://s3-us-west-2.amazonaws.com/allennlp/datasets/quora-question-paraphrase/test.tsv
...
2018-08-21 09:34:18,226 - INFO - allennlp.commands.evaluate - Metrics:
2018-08-21 09:34:18,226 - INFO - allennlp.commands.evaluate - accuracy: 0.893
```
